web/recv.js

```diff
@@ -0,0 +1,401 @@
+            video.srcObject = localMediaStream;
+          } else {
+            video.src = URL.createObjectURL(localMediaStream); //deprecated
+          }
+          video.play();
+          video.requestVideoFrameCallback(Recv.on_frame);
+        })
+        .catch(err => {
+          console.error(`OH NO!!!!`, err);
+        });
+    },
+
+    restart_video: function () {
+      console.log("restart video?");
+      if (_video) {
+        return;
+      }
+      if (!('srcObject' in _video)) {
+        console.log("can't restart video? :|")
+        return;
+      }
+      // ios cancels the feed sometimes. Restart it...?
+      // check if localMediaStream.muted after a while???
+      if (_video.srcObject.muted) {
+        Recv.init_video(_video);
+      }
+    },
+
+    download_bytes: function (buff, name) {
+      var blob = new Blob([buff], { type: 'application/octet-stream' });
+      var bloburl = window.URL.createObjectURL(blob);
+      _downloadHelper(name, bloburl);
+      setTimeout(function () {
+        return window.URL.revokeObjectURL(bloburl);
+      }, 1000);
+    },
+
+    on_decode: function (wid, data) {
+      console.log('Main thread received message from worker' + wid + ':', data);
+      // if extract but no bytes, log extract counter
+      if (data.nodata) {
+        _recentExtract = _counter;
+      }
+      if (data.res) {
+        Recv.set_HTML("t" + wid, "avg red " + wid + " is " + data.res);
+        return;
+      }
+      if (data.ready) {
+        if (_workerReady)
+          _workerReady();
+        return;
+      }
+
+      // should be a decode with some bytes, so set decodecounter
+      _recentDecode = _counter;
+
+      const buff = new Uint8Array(data);
+      Recv.set_HTML("t" + wid, "len() is " + buff.length + ", buff: " + buff);
+      Sink.on_decode(buff);
+    },
+
+    on_frame: function (now, metadata) {
+      //console.log("on frame");
+      // https://developer.mozilla.org/en-US/docs/Web/API/VideoFrame
+
+      _counter += 1;
+      if (_workers.length == 0)
+        return;
+      if (_nextWorker >= _workers.length)
+        _nextWorker = 0;
+
+      // piggyback off this call to make sure our visual state is correct
+      Recv.update_visual_state();
+
+      var vf = undefined;
+      try {
+        vf = new VideoFrame(_video, { timestamp: now });
+        const width = vf.displayWidth;
+        const height = vf.displayHeight;
+        Recv.set_HTML("errorbox", vf.format, true);
+
+        // try to use the default format, but only if we can decode it...
+        let vfparams = {};
+        if (!_supportedFormats.includes(vf.format)) {
+          vfparams.format = "RGBA";
+        }
+        const size = vf.allocationSize(vfparams);
+        const buff = new Uint8Array(size);
+        vf.copyTo(buff, vfparams);
+
+        let format = vfparams.format || vf.format;
+        if (format == "RGBA" && size != width * height * 4) {
+          format = vf.format; //fallback
```

Comment:
It turns out that on ios, you can ask it for RGBA and it'll happily give you back NV12.